### PR TITLE
Add benchmarks for schema generation with recursive models

### LINKED ISSUE #9711: Expand breadth of benchmarks
We have a robust benchmark test suite in `pydantic-core`, but are lacking similar coverage in `pydantic`.

Specifically, I'm interested in expanding this test suite in relation to our core schema build times, as we're looking to improve performance there in the short term. Secondarily, I'm interested in better benchmarking for our json schema generation. Benchmarks for validation and serialization are great, but we've already got pretty good coverage in `pydantic-core` for those, so I'm less concerned about such benchmarks.

This is a great first issue - you can help write tests that test a variety of use cases for `pydantic`, without having to delve into the internals of `pydantic` code to fix bugs or add new features.

Here are some cases for which I think we should expand our suite of benchmarks (assume these are for core schema generation, but you can write related tests for any of the above mentioned categories). As in, you could read the first bullet as "we need to write some tests showcasing how long schema building takes for models with custom validators".

- [x] custom validators (field and model, also before, wrap, after, and plain)
- [x] tagged and untagged unions
- [x] stdlib types
- [x] second class Pydantic types like `SecretStr` and `PositiveInt`
- [x] custom serializers (field and model, also before, wrap, after, and plain)
- [x] recursive model definitions -- need much more robust testing when defs / refs are involved
- [ ] third party, custom types

I wouldn't say these are necessary to close this issue, but one thing for more involved contributors to look at would be benchmarking specific parts of our core schema generation process in `pydantic`, such as:

- [ ] annotation application
- [ ] discriminator application
- [ ] schema cleaning (including defs / refs simplification)
- [ ] schema walking
- [ ] field acquisition / setup in `ModelMetaClass.__new__`

### PR DESCRIPTION
**Pull request by AdolfoVillalobos**
## Change Summary

A suite of benchmarks for schema generation was added when recursive models were present.

Some of the cases considered:
- Simple self-referencing models (with and without generics)
- Complex recursive models (with and without generics)
- Recursive models with discriminated unions (`BaseModel`, data classes, `TypedDict`)
- Deeply nested recursive models

This should be a good starting point. I'm open to edits, suggestions, and other use cases.

## Related issue number

fix #9711 

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: tests/benchmarks/test_model_schema_generation.py [hunk 1]
```diff
@@ -307,3 +307,115 @@ class PydanticTypeModel(DeferredModel):
...
+
+    @dataclass
+    class Bar:
+        type: Literal['bar']
+
+    @dataclass(frozen=True, kw_only=True, config={'defer_build': True})
+    class Foobar:
+        value: Union[Foo, Bar] = Field(discriminator='type')
+
+    benchmark(lambda: rebuild_dataclass(Foobar, force=True, _types_namespace={}))
```
Comment (Viicos) [inline tests/benchmarks/test_model_schema_generation.py:407]: I think we can remove these ones, I don't expect any schema generation changes with Pydantic models

**Review by Viicos — COMMENTED** [review]
Thanks! Could we add these in a new benchmark file?

File: tests/benchmarks/test_model_schema_generation.py (hunk 1 above)
Comment (AdolfoVillalobos) [reply tests/benchmarks/test_model_schema_generation.py:407]: That makes sense. Thanks for the feedback!

**Comment by AdolfoVillalobos** [comment]
> Thanks! Could we add these in a new benchmark file?

Sure!

File: tests/benchmarks/shared.py [hunk 2]
```diff
@@ -231,3 +231,11 @@ class Foo:
...
     PastDatetime,
 ]
+
+
+class DeferredModel(BaseModel):
+    model_config = {'defer_build': True}
+
+
+def rebuild_model(model: Type[BaseModel]) -> None:
+    model.model_rebuild(force=True, _types_namespace={})
```
Comment (sydney-runkle) [inline tests/benchmarks/shared.py:241]: This is fine bc it's working now, but if we do more complex model rebuilding, we might need to remove this types namespace specification.

**Review by sydney-runkle — APPROVED** [review]
Nice work! Thanks for all of these :)

**Pull request merged**